doc/developer/platform/formalism.md

```diff
@@ -194,3 +199,111 @@ A view introduces *constraints* on the capabilities and frontiers of its input a
+
+All commands are serialized through the timeline.
+A timeline is not a complicated concurrent object.
+
+The timeline timestamps do not need to strictly increase, and any events that have the same timestamp are *concurrent*.
+Concurrent events (at the same timestamp) first modify collections and then read collections.
+All writes at a timestamp are visible to all reads at the same timestamp.
+
+SELECT statements observe all data mutations up through and including their timestamp.
+For this reason, we strictly advance the timestamp for each write that occurs after a read, to ensure that it the write is not visible to the read.
+ADAPTER uses `UpdateAndDowngrade` in response to the first read after a write, to ensure that prior writes are readable and to strictly advance the write frontier.
+
+Some DML operations, like UPDATE and DELETE, require a read-write transaction which prevents other writes from intervening.
+In these and other non-trivial cases, we rely on the total order of timestamps to provide the apparent total order of system execution.
+
+---
+
+Several commands support an optional `AS OF <time>` clause, which instructs the Adapter to use a specific query `time`, if valid.
+The presence of this clause opts a command out of the timeline reasoning: it is neither constrainted by nor does it constrain timestamp selection for other commands.
+The command will observe all writes up through `time` (if it reads) and be visible by all reads from `time` onward (if it writes).
+
+---
+
+## Sources
+
+Materialize supports a `CREATE SOURCE` command that binds to a user-specified name a `GlobalId` produced by the Storage layer in response to a `Create` command.
+A created source has an initial read capability and write capability, which may not be at the beginning of time.
+Two sources created with the same arguments are not guaranteed to have the same contents at the same time (said differently: Materialize uses *nominal* rather than *structural* equivalence for sources).
+
+Sources remain active until a `DROP SOURCE` command is received, at which point the Adapter layer drops its read capability.
+
+One common specialization of source is the "table".
+The `CREATE TABLE` command introduces a new source that is not automatically populated by an external source, and is instead populated by `UpdateAndDowngrade` commands.
+The Adapter layer may use a write-ahead log to durably coalesce multiple writes at the same timestamp, as the `UpdateAndDowngrade` command does not otherwise allow this.
+The `DROP TABLE` command drops both the read and write capabilities for the source.
+
+## Indexes
+
+Materialize supports a `CREATE INDEX` command that results in a dataflow that computes and maintains the contents of the index's subject, in indexed representation.
+The `CREATE INDEX` command can be applied to any read-only view whose definition avoids certain non-deterministic functions (e.g. `now()`, `rand()`, environment variables).
```

Comment:
`CREATE INDEX` can also be applied to tables, which aren't read-only views. I also thought we could `CREATE INDEX` on sources directly? "read-only view" confuses me, or maybe I'm just confused.

Reply:
All I really meant by this is that it can't be a SELECT that does weird stuff in it, like idk SELECT .. INTO. I'll try and find better language (the intent is that indexed tables and sources is the view `SELECT * FROM ...`, but .. I confused you so it's a bug).

Reply:
Should be fixed up (changed "read-only view" to "collection").